src/varname.jl

```diff
@@ -231,37 +359,76 @@ julia> @varname(x).indexing
 ()
 
 julia> @varname(x[1]).indexing
-((1,),)
+(@lens _[1])
 
 julia> @varname(x[:, 1]).indexing
-((Colon(), 1),)
+(@lens _[Colon(), 1])
 
 julia> @varname(x[:, 1][2]).indexing
-((Colon(), 1), (2,))
+(@lens _[Colon(), 1][2])
 
 julia> @varname(x[1,2][1+5][45][3]).indexing
-((1, 2), (6,), (45,), (3,))
+(@lens _[1, 2][6][45][3])
 ```
 
 !!! compat "Julia 1.5"
     Using `begin` in an indexing expression to refer to the first index requires at least
     Julia 1.5.
 """
-macro varname(expr::Union{Expr, Symbol})
-    return esc(varname(expr))
+macro varname(expr::Union{Expr, Symbol}, concretize=false)
+    return varname(expr, concretize)
 end
 
-varname(sym::Symbol) = :($(AbstractPPL.VarName){$(QuoteNode(sym))}())
-function varname(expr::Expr)
-    if Meta.isexpr(expr, :ref)
-        sym, inds = vsym(expr), vinds(expr)
-        return :($(AbstractPPL.VarName){$(QuoteNode(sym))}($inds))
+varname(sym::Symbol, concretize=false) = :($(AbstractPPL.VarName){$(QuoteNode(sym))}())
+function varname(expr::Expr, concretize=false)
+    if Meta.isexpr(expr, :ref) || Meta.isexpr(expr, :.)
+        sym = vsym(expr)
+        # Convert `expr` into something `lensmacro` can parse.
+        inds = Setfield.lensmacro(identity, replace_basesym(expr, :_))
+
+        # TODO: Can we do better, i.e. only check if we have `DynamicLens`?
+        return if concretize
+            :($(AbstractPPL.concretize)($(AbstractPPL.VarName){$(QuoteNode(sym))}($inds), $(esc(sym))))
+        else
+            :($(AbstractPPL.VarName){$(QuoteNode(sym))}($inds))
+        end
     else
         error("Malformed variable name $(expr)!")
     end
 end
 
 
+"""
+    replace_basesym(expr, sub::Symbol)
+
+Return `expr` with the base symbol replaced, e.g.
+`:(x[1].a[end])` results in `:(\$(sub).[1].a[end])`
+
+# Example
+```jldoctest
+julia> AbstractPPL.replace_basesym(:(x[1].a[end][:]), :_)
+:(((_[1]).a[end])[:])
+
+julia> AbstractPPL.replace_basesym(:(x), :_)
+:x
+
+julia> AbstractPPL.replace_basesym(:(1), :_)
+1
+```
+
+"""
+replace_basesym(x, sub::Symbol) = x
```

Comment:
https://github.com/jw3126/Setfield.jl/blob/544346fce31eb904a951eb651e69fe0fe24490b2/src/sugar.jl#L69

Also, perhaps we can reuse [parse_obj_lens](https://github.com/jw3126/Setfield.jl/blob/544346fce31eb904a951eb651e69fe0fe24490b2/src/sugar.jl#L133).

Reply:
`replace_underscore` is the opposite of what we want here:) And we don't really want to do something like `postwalk` since we're _only_ interested in replacing the base symbol.

Reply:
`parse_obj_lens` is a great call though!